console/v1/types_console_sample.go

```diff
@@ -0,0 +1,166 @@
+package v1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ConsoleSample is an extension to customizing OpenShift web console by adding samples.
+//
+// Compatibility level 2: Stable within a major release for a minimum of 9 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=2
+type ConsoleSample struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata"`
+
+	// +kubebuilder:validation:Required
+	// +required
+	Spec ConsoleSampleSpec `json:"spec"`
+}
+
+// ConsoleSampleSpec is the desired sample for the web console.
+// Samples will appear with their title, descriptions and a badge in a samples catalog.
+type ConsoleSampleSpec struct {
+	// title is the display name of the sample.
+	//
+	// It is required and must be no more than 50 characters in length.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=50
+	// +required
+	Title string `json:"title"`
+
+	// abstract is a one-liner that describes the sample.
+	//
+	// It is required and is shown on the sample card tile below the title and provider.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MaxLength=100
+	// +required
+	Abstract string `json:"abstract"`
+
+	// description of the sample. A README.md-like content for additional information,
+	// links, pre and post conditions and instrucations.
+	//
+	// It is required and is shown on the details page. It is rendered as Markdown
+	// so that it can contain line breaks, links, and other simple formatting.
+	// +kubebuilder:validation:Required
+	// +required
+	Description string `json:"description"`
+
+	// icon is an optional base64 encoded image that will be displayed beside the sample title.
+	// The icon should be an vector image for easy scaling.
+	// It will be shown on a white background (also when the dark theme is used)
+	// and in different sizes (currently not bigger then 40x100px).
+	// Most icons will be quadratic but the web console ensures that all aspect ratios work.
```

Comment:
Not sure i follow this, can you expand?

Reply:
Updated to:

4. `icon` is an optional base64 encoded image and shown beside the sample title.
   The format must follow the [data: URL format](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs).

   `data:[<mediatype>][;base64],<base64 encoded image>`

   For example:

   `data:image;base64,` plus the base64 encoded SVG image.

   Vector images can also be used. SVG icons must start with:

   `data:image/svg+xml;base64,` plus the base64 encoded SVG image.

   All sample catalog icons will be shown on a white background (also when the dark theme is used).
   The web console ensures that different aspect radios work correctly.
   Currently, the surface of the icon is at most 40x100px.

Wdyt?

Reply:
Yeah I think that's good. Seems like there's a fixed format here, can we add a regex to validate it at all?